zkvm/src/signer/mod.rs

```diff
@@ -0,0 +1,229 @@
+#![allow(non_snake_case)]
+
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand;
+
+// Modules for signing protocol
+
+pub mod prover;
+
+#[derive(Clone)]
+pub struct PrivKey(Scalar);
+
+#[derive(Clone)]
+pub struct PubKey(RistrettoPoint);
+
+pub struct MultiKey(Vec<PubKey>); // TODO: also include Option<Scalar> for signing key?
+
+#[derive(Clone)]
+pub struct PubKeyHash(Scalar);
+
+#[derive(Debug, Clone)]
+pub struct Signature {
+    s: Scalar,
+    R: RistrettoPoint,
+}
+
+#[derive(Clone)]
+pub struct Shared {
+    G: RistrettoPoint,
+    transcript: Transcript,
+    // X_agg = sum_i ( a_i * X_i )
+    X_agg: PubKey,
+    // L = H(X_1 || X_2 || ... || X_n)
+    L: PubKeyHash,
+    // message being signed
+    m: Vec<u8>,
+}
+
+impl MultiKey {
+    pub fn aggregate(&self, transcript: &mut Transcript) -> (PubKey, PubKeyHash) {
+        // L = H(X_1 || X_2 || ... || X_n)
+        let mut L_transcript = transcript.clone();
+        for X_i in &self.0 {
+            L_transcript.commit_point(b"X_i.L", &X_i.0.compress());
+        }
+        let L = L_transcript.challenge_scalar(b"L");
+
+        // X = sum_i ( a_i * X_i )
+        // a_i = H(L, X_i)
+        let mut X = RistrettoPoint::default();
+        for X_i in &self.0 {
+            let mut a_i_transcript = transcript.clone();
+            a_i_transcript.commit_scalar(b"L", &L);
+            a_i_transcript.commit_point(b"X_i", &X_i.0.compress());
+            let a_i = a_i_transcript.challenge_scalar(b"a_i");
+            X = X + a_i * X_i.0;
+        }
+
+        (PubKey(X), PubKeyHash(L))
+    }
+}
+
+impl Signature {
+    pub fn verify(&self, shared: Shared) -> bool {
```

Comment:
We can safely hardcode ristretto basepoint and pass &mut Transcript here. Then, the API would hardly be different from this: https://github.com/interstellar/slingshot/blob/main/zkvm/src/signature.rs#L28

That is, we also pass in compressed pubkey and R, so we don't need to re-compress them. And decompression is deferred to PointOp.

Note: the current `verify_aggregated` API in zkvm is the poor attempt at multi-message API - please ignore the part about multiple keys.

Reply:
>We can safely hardcode ristretto basepoint and pass &mut Transcript here.
I mean, no need for `Shared`.

Reply:
I mentioned this in the other comment, but the transcript for the aggregation and signer and verifier all have to be the exact same, or else it won't verify properly.

So, if we are hardcoding the transcript into the aggregation, then we should hardcode the transcript into the verify function (instead of passing in &mut Transcript). Which should it be... all hardcoded, or all passed in?

Reply:
Resolved - we are hardcoding the transcript for pubkey aggregation, and passing in the transcript for nonce commitment generation, factor_for_key generation etc (basically, for prover and verifier use)